src/opencl_rawsha512_gpl_fmt_plug.c

```diff
@@ -162,33 +196,49 @@ static uint64_t *crypt_one_x(int index) {
 }
 
 /* ------- Create and destroy necessary objects ------- */
-static void create_mask_buffers()
+static void set_kernel_args()
 {
-	release_mask_buffers();
+	//Set the preparation kernel arguments (cleans and prepares the device memory)
+	CLKERNELARGx(prepare_kernel, 0, mask_int_cand.num_int_cand);
+	CLKERNELARGx(prepare_kernel, 1, dev_cracks);
 
-	saved_bitmap = (uint32_t *)
-		mem_alloc((bitmap_size / 32 + 1) * sizeof(uint32_t));
-	buffer_bitmap = clCreateBuffer(context[gpu_id], CL_MEM_WRITE_ONLY,
-		(bitmap_size / 32 + 1) * sizeof(uint32_t), NULL, &ret_code);
-	HANDLE_CLERROR(ret_code, "Error creating buffer argument buffer_bitmap");
+	//Set the crypt kernel arguments (runs sha512_block())
+	CLKERNELARGx(crypt_kernel, 0, salt_buffer);
+	CLKERNELARGx(crypt_kernel, 1, pinned_plaintext);
+	CLKERNELARGx(crypt_kernel, 2, pinned_saved_idx);
+	CLKERNELARGx(crypt_kernel, 3, pinned_int_key_loc);
+	CLKERNELARGx(crypt_kernel, 4, dev_int_keys);
+	CLKERNELARGx(crypt_kernel, 5, mask_int_cand.num_int_cand);
+	CLKERNELARGx(crypt_kernel, 6, dev_cracks);
+}
 
+static void set_kernel_bitmap()
+{
 	//Set crypt kernel arguments
-	HANDLE_CLERROR(clSetKernelArg(crypt_kernel, 7, sizeof(buffer_bitmap),
-	                              (void *)&buffer_bitmap), "Error setting argument 7");
+	CLKERNELARGx(crypt_kernel, 7, dev_bitmap);
 }
 
-static void release_mask_buffers()
+static void create_bitmap_buffers()
 {
-	MEM_FREE(saved_bitmap);
+	release_bitmap_buffers();
+
+	bitmap = (uint32_t *)
+		mem_alloc((bitmap_size / 32 + 1) * sizeof(uint32_t));
+	opencl_create_buf(&dev_bitmap,
+	    CL_MEM_WRITE_ONLY, (bitmap_size / 32 + 1) * sizeof(uint32_t), CL_DONT_KEEP_TRACK);
```

Comment:
Shouldn't most our functions (or new alternatives to them) cover both?

Something like:
```c
opencl_create_bufs(&bitmap, &dev_bitmap, RO, (bitmap_size / 32 + 1) * sizeof(uint32_t), CL_NO_TRACK);
```
This would cover lines 225-228: Do a malloc for `bitmap` and an OpenCL create for `dev_bitmap` in a single call. With future means for great debugging assertions.

Reply:
Done.

Reply:
We have a problem here. At least one: salt. We usually don't `malloc()` salt.

Reply:
So we need alternative functions: Ones that only allocates a device buffer, and ones that allocate a host buffer and a device buffer with same size.

I guess the `_map` functions can always do both though? Or maybe they already do.

Reply:
It may seem overkill but in my head it will simplify 99% of the formats so I think we should really consider these things. That being said, I'm just brainstorming.

Reply:
If you think that `MEM_FREE()` is going to happen automagically, it looks good.

Reply:
Absolutely. Perhaps we can call this `opencl_create_buf_pair`.  On call to `opencl_release_buf_pairs()` - or whatever we'll call this - we'd release both.

Reply:
Now, we could make these `_pair` functions yet another set of similar functions, or we could do half of them with macros. I'm still merely brainstorming. It could go something like this (just from head):
```c
cl_mem opencl_create_buf_pair(void **host_buf cl_mem *buffer_obj, cl_mem_flags flags, size_t size, int keep_track)
{
	cl_int err_create_buffer;

	if (*(buffer_obj))
#ifdef DEBUG
		error_msg("** %s:%d Memory leak (device buffer)\n", __FILE__, __LINE__);
#else
		opencl_release_buf(buffer_obj);
#endif

	if (**(host_buf))
#ifdef DEBUG
		error_msg("** %s:%d Memory leak (host buffer)\n", __FILE__, __LINE__);
#else
		MEM_FREE(**host_buf);
#endif

	*(buffer_obj) = clCreateBuffer(context[gpu_id], flags, size, NULL, &err_create_buffer);

	if (err_create_buffer != CL_SUCCESS) {
		*(buffer_obj) = NULL;
		HANDLE_CLERROR(err_create_buffer, "Error allocating GPU memory");
	} else
		if (keep_track)
			buf_list_add(cl_buffer_list, buffer_obj);

	if (*host_buf)
		*(host_buf) = mem_alloc(size);

	return *(buffer_obj); /* or should be drop this and just be void? Or return an error code? The buffer was in the parameters */
 }

/* Version for device-side only */
#define opencl_create_buf(buffer_obj, flags, size, keep_track) \
	opencl_create_buf(NULL, buffer_obj, flags, size, keep_track)
```

Reply:
BTW since these are no longer macros, we lose the value of `__FILE__, __LINE__`. We could fix that by wrapping everything with a master macro much like `error_msg()` et al.

Reply:
If it prints `opencl_common.c line 3007` it is ok to me. The same failure/message can happen in some places, the message will be useful. Anyone can say something about macroization?

Reply:
It's easy really.

If the real function is
```c
cl_mem opencl_create_buf_pair(void **host_buf, cl_mem *buffer_obj, cl_mem_flags flags, size_t size, int keep_track)
```
we modify it to be
```c
cl_mem real_opencl_create_buf_pair(void **host_buf, cl_mem *buffer_obj, cl_mem_flags flags, size_t size, int keep_track, char *file, int line)```
and use a macro for calling it
```c
#define opencl_create_buf_pair(host_buf buffer_obj, flags, size, keep_track) \
    real_cl_mem opencl_create_buf_pair(host_buf, buffer_obj, flags, size, keep_track, __FILE__, __LINE__)
```